### DISCUSSION TITLE
ACIA interrupt mode 2

### DISCUSSION THREAD

**Discussion by jpelletier** [wwarthen/RomWBW#297] [Q&A] [answered]
In HBIOS/acia.asm, the code generates an error: "acia device does not support intmode 2".
Why exactly? My Z80 system supports interrupt mode 2 and can generate a vector for 24 interrupts.

**Comment by wwarthen** [comment]
The ACIA has no inherent support for placing a Z80 mode 2 interrupt vector on the bus like Zilog peripheral chips such as PIO, CIO, CTC, etc.  As such it cannot create a mode 2 interrupt without some help.  I guess it would be more appropriate to say that the ACIA does not support intmode 2 by itself.

Does your system have a mechanism for generating a mode 2 interrupt vector from  the ACIA?  If so, then it should be possible to support it.

Thanks,

Wayne

**Reply by jpelletier** [reply]
I made my own interrupt controller (combined with other logic) with a CPLD. It's cascadable and have a register to enable/disable interrupts. I just finished the debugging phase a few week ago.

The source code is written in AHDL to use in Altera chips.
https://sourceforge.net/projects/monsputer-bios/files/CPLD/CPM/IO6850.tdf
I cascaded it into another CPLD from Atmel. The source is written in WinCUPL.
https://sourceforge.net/projects/monsputer-bios/files/CPLD/Z80SERIAL.PLD

**Reply by wwarthen** [reply, marked as answer]
Sounds like a nice solution.  I don't know enough about AHDL or CPLD to fully comprehend your design.  If it is is allowing an ACIA to generate the IM2 interrupt vectors that the Z80 wants, then you could certainly modify the ACIA driver to remove the IM2 constraint.  The IM2 constraint is in the ACIA driver because interrupt mode 2 will fail for all other ACIA implementations.

As far as I know, this is a one-off design.  If it is being distributed to a wider audience, we could consider incorporating support for it in mainstream RomWBW.

Thanks,

Wayne
